server/registry/actions_apis.go

```diff
@@ -202,29 +185,29 @@ func (s *RegistryServer) UpdateApi(ctx context.Context, req *rpc.UpdateApiReques
 	}
 
 	if req.GetAllowMissing() {
-		before := time.Now()
-		// Prevent a race condition that can occur when two updates are made
-		// to the same non-existent resource. The db.Get...() call returns
-		// NotFound for both updates, and after one creates the resource,
-		// the other creation fails. The lock() prevents this by serializing
-		// the get and create operations. Future updates could improve this
-		// with improvements closer to the database level.
-		updateApiMutex.Lock()
-		defer updateApiMutex.Unlock()
-		log.Debugf(ctx, "Acquired lock after blocking for %v", time.Since(before))
+		// Speculatively create the api. If it already exists, fall through to the update below.
+		api, err := s.createApi(ctx, name, req.GetApi())
+		if err == nil {
+			return api, nil
+		} else if status.Code(err) != codes.AlreadyExists {
+			return nil, err
+		}
+	}
+
+	db, err := s.getStorageClient(ctx)
+	if err != nil {
+		return nil, status.Error(codes.Unavailable, err.Error())
 	}
+	defer db.Close()
 
+	// TODO: the Get/Update/Save calls below should be a transaction.
```

Comment:
Is there an issue filed to address this?

Reply:
#623